nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java

```diff
@@ -0,0 +1,232 @@
+        if (stateMap.getVersion() == -1L || stateMap.get(CURRENT_TIMESTAMP) == null || stateMap.get(CURRENT_KEY_PREFIX) == null) {
+            currentTimestamp = 0L;
+            currentKeys = new HashSet<>();
+        } else {
+            currentTimestamp = Long.parseLong(stateMap.get(CURRENT_TIMESTAMP));
+            currentKeys = extractKeys(stateMap);
+        }
+    }
+
+    private void persistState(final ProcessContext context) {
+        Map<String, String> state = new HashMap<>();
+        state.put(CURRENT_TIMESTAMP, String.valueOf(currentTimestamp));
+        int i = 0;
+        for (String key : currentKeys) {
+            state.put(CURRENT_KEY_PREFIX+i, key);
+        }
+        try {
+            context.getStateManager().setState(state, Scope.CLUSTER);
+        } catch (IOException ioe) {
+            getLogger().error("Failed to save cluster-wide state. If NiFi is restarted, data duplication may occur", ioe);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        try {
+            restoreState(context);
+        } catch (IOException ioe) {
+            getLogger().error("Failed to restore processor state; yielding", ioe);
+            context.yield();
+            return;
+        }
+
+        final long startNanos = System.nanoTime();
+        final String bucket = context.getProperty(BUCKET).evaluateAttributeExpressions().getValue();
+
+        final AmazonS3 client = getClient();
+        int listCount = 0;
+        long maxTimestamp = 0L;
+        String delimiter = context.getProperty(DELIMITER).getValue();
+        String prefix = context.getProperty(PREFIX).evaluateAttributeExpressions().getValue();
+
+        ListObjectsRequest listObjectsRequest = new ListObjectsRequest().withBucketName(bucket);
+        if (delimiter != null && !delimiter.isEmpty()) {
+            listObjectsRequest.setDelimiter(delimiter);
+        }
+        if (prefix != null && !prefix.isEmpty()) {
+            listObjectsRequest.setPrefix(prefix);
+        }
+
+        ObjectListing objectListing;
+        do {
+            objectListing = client.listObjects(listObjectsRequest);
+            for (S3ObjectSummary objectSummary : objectListing.getObjectSummaries()) {
+                long lastModified = objectSummary.getLastModified().getTime();
+                if (lastModified < currentTimestamp
+                        || lastModified == currentTimestamp && currentKeys.contains(objectSummary.getKey())) {
+                    continue;
+                }
+
+                // Create the attributes
+                final Map<String, String> attributes = new HashMap<>();
+                attributes.put(CoreAttributes.FILENAME.key(), objectSummary.getKey());
+                attributes.put("s3.bucket", objectSummary.getBucketName());
+                attributes.put("s3.owner", objectSummary.getOwner().getId());
```

Comment:
Good catch - looks like `getOwner()` can <a href="http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/S3ObjectSummary.html#getOwner()">return null</a>. Should be fixed now.